third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -293,30 +416,60 @@ impl InterpreterImpl {
-                    let ty_args = resolver
-                        .instantiate_generic_function(
-                            Some(gas_meter),
-                            idx,
-                            current_frame.function.ty_args(),
-                        )
-                        .map_err(|e| set_err_info!(current_frame, e))?;
-                    let function = resolver
-                        .build_loaded_function_from_instantiation_and_ty_args(idx, ty_args)
-                        .map_err(|e| self.set_location(e))?;
+                    let (function, frame_cache) = if RTCaches::call_tree_cache_enabled() {
+                        let current_frame_cache = &mut *current_frame.ty_cache.borrow_mut();
 
-                    if RTTCheck::should_perform_checks() {
-                        self.check_friend_or_private_call(&current_frame.function, &function)?;
-                    }
+                        match current_frame_cache.generic_sub_frame_cache.entry(idx) {
+                            std::collections::btree_map::Entry::Occupied(entry) => {
+                                let entry = entry.get();
+                                (Rc::clone(&entry.0), Rc::clone(&entry.1))
+                            },
+                            std::collections::btree_map::Entry::Vacant(entry) => {
+                                let function =
+                                    Rc::<LoadedFunction>::new(self.load_generic_function(
+                                        &resolver,
+                                        &current_frame,
+                                        gas_meter,
+                                        idx,
+                                    )?);
+                                let frame_cache = Rc::new(RefCell::new(Default::default()));
+
+                                if RTCaches::per_instruction_cache_enabled()
+                                    && !function.is_native()
+                                {
+                                    let f_cache: &mut FrameTypeCache =
+                                        &mut frame_cache.borrow_mut();
+                                    f_cache
+                                        .per_instruction_cache
+                                        .resize(function.code_size(), PerInstructionCache::Nothing);
+                                }
+
+                                entry.insert((Rc::clone(&function), Rc::clone(&frame_cache)));
+                                (function, frame_cache)
+                            },
+                        }
+                    } else {
+                        let function = Rc::<LoadedFunction>::new(self.load_generic_function(
+                            &resolver,
+                            &current_frame,
+                            gas_meter,
+                            idx,
+                        )?);
+                        let frame_cache = Rc::new(RefCell::new(Default::default()));
```

Comment:
not: explicit type instead of default. Btw, here you do not resize? Why? If you have a separate callback for `per_instruction_cache_enabled` I would expect it to be possible to have it as true, but `call_tree_cache_enabled` as false. Then the code can panic? 

I think here we need to either check this, or just have one function to enable all three?